### DISCUSSION TITLE
The garbage collector is trying to clean up non-checked-in connection - AsyncSession with Connection Pooling

### DISCUSSION THREAD

**Discussion by jcheso** [sqlalchemy/sqlalchemy#11178] [Usage Questions] [answered]
Hi all,

I am encountering the following error on my production server, which I am not able to reproduce locally and am having a hard time tracking down the cause. Any guidance on how to troubleshoot it further or what may be causing it would be greatly appreciated. I believe it may be related to using `AsyncSession`, as I recently updated from `Session` to `AsyncSession` (along with package version updates), and the error started occurring after that.

​```
The garbage collector is trying to clean up non-checked-in connection <AdaptedConnection <asyncpg.connection.Connection object at 0x3dfdd19ee4d0>>, which will be terminated. Please ensure that SQLAlchemy pooled connections are returned to the pool explicitly, either by calling ``close()`` or by using appropriate context managers to manage their lifecycle.
​```

I am running a FastAPI server (runs in a Docker image with Gunicorn/Uvicorn on GCP with Cloud SQL Postgres), using asyncpg and SQLAlchemy for my database ORM with the following versions:

- asyncpg = "^0.29.0"
- sqlalchemy = "^2.0.28"
- fastapi = "^0.110.0"
- uvicorn = "^0.29.0"

I have a `DatabaseSessionManager` class that manages the creation and closing of sessions:

```python
class DatabaseSessionManager:
    def __init__(self, host: str, engine_kwargs: dict[str, Any] = {}):
        self._engine = create_async_engine(
            host,
            pool_size=70,
            max_overflow=30,
            pool_timeout=30,
            pool_recycle=1800,
            echo_pool=True,
            **engine_kwargs,
        )
        self._sessionmaker = async_sessionmaker(autocommit=False, bind=self._engine)

    async def close(self):
        if self._engine is None:
            raise Exception("DatabaseSessionManager is not initialized")
        logger.debug("Disposing engine")
        await self._engine.dispose()

        self._engine = None
        self._sessionmaker = None

    @contextlib.asynccontextmanager
    async def connect(
        self,
    ) -> AsyncIterator[AsyncConnection]:
        if self._engine is None:
            raise Exception("DatabaseSessionManager is not initialized")

        async with self._engine.begin() as connection:
            try:
                yield connection
            except:
                await connection.rollback()
                raise
            finally:
                await connection.close()

    @contextlib.asynccontextmanager
    async def session(
        self,
    ) -> AsyncIterator[AsyncSession]:
        if self._sessionmaker is None:
            raise Exception("DatabaseSessionManager is not initialized")

        session = self._sessionmaker()
        try:
            yield session
        except:
            await session.rollback()
            raise
        finally:
            await session.close()

```

I then use FastAPI dependency injection to create an injectable `DBSessionDep`:

```python
async def get_db_session():
    async with session_manager.session() as session:
        yield session

DBSessionDep = Annotated[AsyncSession, Depends(get_db_session)]
```

This `DBSessionDep` is the only way I ever access the database, and it is injected into each endpoint. I am under the impression that it should always call the `close()` method on the session before the garbage collector (GC) runs, but that doesn't seem to be occurring.

I have also ensured that I am not manually closing the session outside of the `session` context manager and that I have appropriate exception handling in place to gracefully handle any errors and allow the session to be closed.

If I've omitted any key details, please let me know. Apologies that I cannot provide a reproducible sample easily, as it is very sporadic - I've tried logging the session and connection IDs to track down where exactly it is coming from but didn't manage to find any form of pattern.

Any insights or suggestions on how to further diagnose and resolve this issue would be highly appreciated. Thank you in advance for your help!

**Comment by zzzeek** [comment]
hi -

Have you also reached out to FastAPI experts for assistance on this?    there's nothing obviously wrong with any of the code, we get these issues continuously, only with FastAPI, no other frameworks.  Typically this `Depends` construct is often on the scene however, we dont maintain that here or know anything about it.

**Reply by jcheso** [reply]
Thanks, good to know. I thought I would start here in case it was an obvious misconfiguration with SQLAlchemy - I'll head over to FastAPI and check with their community.

**Reply by zzzeek** [reply]
one thing to keep in mind is that nearly always with these "occasional errors in the logs" things, the issue has to do with requests that are cancelled before completing.

**Comment by CaselIT** [comment]
Hi,

Assuming fastapi does the correct thing with `get_db_session` then it this should be ok.
Does uvicorn use the default asyncio loop or uses instead a different implementation? If so to may make sense trying with the default loop

**Reply by jcheso** [reply]
Thanks for checking - I am using  the default asyncio loop: `<class 'asyncio.unix_events._UnixSelectorEventLoop'>`

**Reply by CaselIT** [reply, marked as answer]
You may also try to use this parameter https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.close_resets_only setting it to false will raise an error if you try to use the session after it's closed. Maybe the issue is that you continue using the session after close was called, re-opening it and leaving it open

**Reply by jcheso** [reply]
This helped me track down the error - it appears that when I was injecting a session into an endpoint, then passing it to a background task, the session was closing when the endpoint returned, and was being re-opened by the background task (and never closed). 

Thanks a lot for your help!

**Reply by CaselIT** [reply]
note that the background task should have its own session since a session cannot be used concurrently

**Reply by PREPONDERANCE** [reply]
@jcheso  Hi!

I'd like to know how did you set `close_resets_only` to `False` to  a session created by `async_sessionmaker`.

I tried to set `self._close_state = _SessionCloseState.CLOSED` but no errors are given out. The only messages I receive are the gc warnings.

```Python
async_engine = create_async_engine(settings.DATABASE_URL, echo=False)
async_db_session = async_sessionmaker(
    bind=async_engine, class_=AsyncSession, expire_on_commit=False, autoflush=False
)

class DataBaseSession:
    def __init__(self) -> None:
        self.db = None

    async def __aenter__(self) -> AsyncSession:
        self.db = async_db_session()
        return self.db

    async def __aexit__(self, exc_type, exc_value, traceback):
        await self.db.close()

async def get_db_session():
    async with DataBaseSession() as session:
        yield session
```

**Reply by CaselIT** [reply]
I think you are missing `@asynccontextmanager` on your `get_db_session`

**Reply by PREPONDERANCE** [reply]
But this function serves as the dependency for FastAPI endpoints. I tried adding `@asynccontextmanager` to this function but it compromised other functionalities.

**Reply by CaselIT** [reply]
I don't know what fastapi does here. In plain python you would use that decorator for it to work as you expect 

You can see whee the close is called by placing a log in the `__exit__`

**Reply by PREPONDERANCE** [reply]
Thanks! 

Indeed the async context manager works as expected and considering in FastAPI the background tasks will be called after the endpoint is finished, in which case, reopening the database connection leads to gc warnings, I reset the database connection whenever a background task is required. But the problems remain.

So is there a way that I can debug the SQLAlchemy session as described in the documentation? I find the `close_resets_only` parameter in `Session.__init__` function, but how should I change it when I initialize a session through `async_sessionmaker`?

**Reply by CaselIT** [reply]
```py
async_db_session = async_sessionmaker(
    bind=async_engine, class_=AsyncSession, expire_on_commit=False, autoflush=False
)
```

That you have definined here are session init arguments. If you want to change only a single session you can also provide them when calling async_db_session()

**Reply by PREPONDERANCE** [reply]
Thanks!

At the end of the day, it's the problem with how I handle database connection in the background task.

**Reply by jcheso** [reply]
Hi @PREPONDERANCE, did you manage to figure it out?
From memory I just passed `close_resets_only` into the sessionmaker like so:
```python
async_db_session = async_sessionmaker(
    bind=async_engine, class_=AsyncSession, expire_on_commit=False, autoflush=False, close_resets_only=False
)
```

**Reply by PREPONDERANCE** [reply]
Yes, I've managed that out with the help of CaselIT. Thanks for your reply!
